src/blog/adaptor-frame.md

```diff
@@ -0,0 +1,315 @@
+
+Note however to be able to access the arguments, the function must know how many arguments are in the stack! The index `2` points to the last argument regardless of how many arguments there are!
+
+The bytecode handler of `Return` will finish by calling the builtin `LeaveInterpreterFrame`. This builtin essentially reads the function object to get the parameter count from the frame, pops the current frame, recovers the frame pointer, saves the return address in a scratch register, pops the arguments according to the parameter count and jumps to the address in the scratch registers.
+
+All this flow is great! But what happens when we call a function with fewer or more arguments than its parameter count? The clever argument/register access will fail and how do we clean up the arguments at the end of the call?
+
+## Arguments adaptor frame
+
+Let’s now call `add42` with fewer and more arguments:
+
+```js
+add42();
+add42(1, 2, 3);
+```
+
+The JS developers between us will know that in the first case, `x` will be assigned `undefined` and the function will return `undefined + 42 = NaN`. In the second case, `x` will be assigned `1` and the function will return `43`, the remaining arguments will be ignored. Note that the caller does not know if that will happen. Even if the caller checks the parameter count, the callee could use the rest parameter or the arguments object to access all the other arguments. Actually, the arguments object can even be accessed outside `add42` in sloppy mode.
+
+If we follow the same steps as before, we will first call the builtin `InterpreterPushArgsThenCall`. It will push the arguments to the stack like so:
+
+![State of the frames after the execution of `InterpreterPushArgsThenCall` builtin.](/_img/adaptor-frame/adaptor-push.svg)
+
+Continuing the same procedure as before, we check if the callee is a function object, get its parameter count and patch the receiver to the global proxy. Eventually we reach `InvokeFunctionCode`.
+
+Here instead of jumping to the `Code` in the callee object. We check that we have a mismatch between argument size and parameter count and jump to `ArgumentsAdaptorTrampoline`.
+
+In this builtin, we build an extra frame, the infamous arguments adaptor frame. Instead of explaining what happens inside the builtin, I will just present you the state of the frame before the builtin calls the callee object’s `Code`. Note that this is a proper `x64 call` (not a `jmp`) and after the execution of the callee we will return to the `ArgumentsAdaptorTrampoline`. This is a contrast with `InvokeFunctionCode` that tailcalls.
```

Comment:
```suggestion
In this builtin, we build an extra frame, the infamous arguments adaptor frame. Instead of explaining what happens inside the builtin, I will just present you the state of the frame before the builtin calls the callee’s `Code`. Note that this is a proper `x64 call` (not a `jmp`) and after the execution of the callee we will return to the `ArgumentsAdaptorTrampoline`. This is a contrast with `InvokeFunctionCode` that tailcalls.
```